polar-core/src/sugar.rs

```diff
@@ -0,0 +1,850 @@
+        permissions.value().as_list().unwrap().is_empty()
+    });
+    let relations_empty = relations.as_ref().map_or(false, |relations| {
+        relations.value().as_dict().unwrap().is_empty()
+    });
+
+    match (roles_empty, permissions_empty, relations_empty) {
+        (true, true, true) => Err(ParseError::IntegerOverflow {
+            loc: resource.offset(),
+            token: format!(
+                "{} namespace contains empty roles, permissions, and relations declarations. \
+                        Please add roles, permissions, and relations or delete the declarations.",
+                resource
+            ),
+        }
+        .into()),
+        (true, true, _) => Err(ParseError::IntegerOverflow {
+            loc: resource.offset(),
+            token: format!(
+                "{} namespace contains empty roles and permissions declarations. \
+                        Please add roles and permissions or delete the declarations.",
+                resource
+            ),
+        }
+        .into()),
+        (true, _, true) => Err(ParseError::IntegerOverflow {
+            loc: resource.offset(),
+            token: format!(
+                "{} namespace contains empty roles and relations declarations. \
+                        Please add roles and relations or delete the declarations.",
+                resource
+            ),
+        }
+        .into()),
+        (_, true, true) => Err(ParseError::IntegerOverflow {
+            loc: resource.offset(),
+            token: format!(
+                "{} namespace contains empty permissions and relations declarations. \
+                        Please add permissions and relations or delete the declarations.",
+                resource
+            ),
+        }
+        .into()),
+        (true, _, _) => Err(ParseError::IntegerOverflow {
+            loc: roles.as_ref().unwrap().offset(),
+            token: format!(
+                "{} namespace contains an empty roles declaration. \
+                        Please add roles or delete the declaration.",
+                resource
+            ),
+        }
+        .into()),
+        (_, true, _) => Err(ParseError::IntegerOverflow {
+            loc: resource.offset(),
+            token: format!(
+                "{} namespace contains an empty permissions declaration. \
+                        Please add permissions or delete the declaration.",
+                resource
+            ),
+        }
+        .into()),
+        (_, _, true) => Err(ParseError::IntegerOverflow {
+            loc: resource.offset(),
+            token: format!(
+                "{} namespace contains an empty relations declaration. \
+                        Please add relations or delete the declaration.",
+                resource
+            ),
+        }
+        .into()),
+        (false, false, false) => Ok(()),
+    }
+}
+
+fn check_all_permissions_involved_in_implications(namespace: &Namespace) -> PolarResult<()> {
```

Comment:
Is it punishment? I think of it more like the Elm / Rust compiler guidance, both of which will fail to compile upon each newly encountered error but will helpfully point out the problem (and sometimes even offer potential solutions).

I think it will feel less like punishment when it's hooked into your IDE and updating as you type as opposed to getting surfaced through a "write some stuff, then try running the program and encounter an error" cycle.

I see what you're saying though... there's a difference between "this is incorrect" and "this is unused and maybe harmful to tidiness but not _wrong_".

Reply:
Yeah I guess it is kinda more like Clippy pointing out needless clones and whatnot, which are warnings but not hard errors.